nimbus-consensus/src/lib.rs

```diff
@@ -195,18 +197,42 @@ where
 		}
 
 		let at = BlockId::Hash(parent.hash());
-		// Get `AuthorFilterAPI` version.
-		let api_version = self.parachain_client.runtime_api()
-			.api_version::<dyn AuthorFilterAPI<B, NimbusId>>(&at)
-			.expect("Runtime api access to not error.");
-
-		if api_version.is_none() {
-			tracing::error!(
-				target: LOG_TARGET, "Could not find `AuthorFilterAPI` version.",
-			);
-			return None;
-		}
-		let api_version = api_version.unwrap();
+
+		// helper function for calling the various runtime apis and versions
+		let prediction_helper = |at, nimbus_id: NimbusId, slot: u32, parent| -> bool {
+
+			let has_nimbus_api = self
+				.parachain_client
+				.runtime_api()
+				.has_api::<dyn NimbusApi<B>>(at)
+				.expect("should be able to dynamically detect the api");
+			
+			if has_nimbus_api {
+				// I understand that this is ambiguous, but I don't understand how to disambiguate.
+				// The compiler's suggestion does not work. See where I called the author filter api below.
+				self.parachain_client.runtime_api().can_author(at, nimbus_id, slot, parent)
+					.expect("NimbusAPI should not return error")
```

Comment:
@librelois I can't figure out how to make this compile.

Reply:
When you call a method implicitly (`var.method()`), the rust compiler will search in the implementations of the `var` type and if it doesn't find it, it will automatically search in the implementations of the deref target, which is not done when calling a method explicitly, so you just need to deref explicitly:

```
use core::ops::Deref as _;
NimbusApi::can_author(self.parachain_client.runtime_api().deref(), at, nimbus_id, slot, parent)
    .expect("NimbusAPI should not return error")
```

Reply:
Thank you for the solution and the explanation!